lib/puppet/util/execution.rb

```diff
@@ -216,28 +228,32 @@ def self.execute(command, options = NoOptionsSpecified)
       elsif Puppet.features.microsoft_windows?
         process_info = execute_windows(*exec_args)
         begin
+          [stdin, stdout, stderr].each {|io| io.close rescue nil}
+          unless options[:squelch]
+            while !reader.eof?
+              output << reader.read
+            end
+            read_succeeded = true
+          end
           exit_status = Puppet::Util::Windows::Process.wait_process(process_info.process_handle)
         ensure
           FFI::WIN32.CloseHandle(process_info.process_handle)
           FFI::WIN32.CloseHandle(process_info.thread_handle)
         end
       end
 
-      [stdin, stdout, stderr].each {|io| io.close rescue nil}
-
-      # read output in if required
-      unless options[:squelch]
-        output = wait_for_output(stdout)
-        Puppet.warning "Could not get output" unless output
+      unless options[:squelch] || read_succeeded
+        Puppet.warning "Could not get output"
+        output = nil
       end
 
       if options[:failonfail] and exit_status != 0
         raise Puppet::ExecutionFailure, "Execution of '#{command_str}' returned #{exit_status}: #{output.strip}"
       end
     ensure
-      if !options[:squelch] && stdout
-        # if we opened a temp file for stdout, we need to clean it up.
-        stdout.close!
+      if !options[:squelch] && reader
+        # if we opened a pipe, we need to clean it up.
+        reader.close
```

Comment:
If an exception is raised between lines 190 and 201 (posix) or 207 (windows), then I don't think we close the `writer`:

```ruby
begin
  reader, writer = IO.pipe unless options[:squelch]
  ...
  process_info = execute_windows(*exec_args) # if this raises
  begin
    [stdin, stdout, stderr].each {|io| io.close rescue nil} # this line is never reached
    ...
   end
ensure
  if !options[:squelch] && reader
    reader.close
  end
end
```

A good test is to try to execute a non-existing program with `squelch = false`. On Windows, I know `execute_windows` will raise, and I think the posix impl behaves the same.